### PR TITLE
Euclidean wheel conv

### LINKED ISSUE #2406: Create transformation matrix for local to per-wheel velocities
We need to convert local velocities to per-wheel-velocities that we can
send to the trinamic motor controllers.

See: http://robocup.mi.fu-berlin.de/buch/omnidrive.pdf
See: `shared/physics.c:shared_physics_speed4ToSpeed3`
See: `shared/physics.c:shared_physics_speed3ToSpeed4`

Currently these matrices are hard coded, and they can continue to be if that makes it easier

But it would be super cool to generate them from the wheel angles. That information is stored in RobotConstants_t (see `2021_robot_constants.c`). Also see https://github.com/UBC-Thunderbots/Software/issues/2099

### Acceptance criteria

- [ ] Figure out local velocity -> per wheel velocity conversion matrix a
- [ ] Figure out per wheel velocity -> local velocity conversion matrix b
- [ ] Create a function that takes local speed, applies matrix a, and returns per wheel velocities
- [ ] Create a function that takes per wheel velocities, applies matrix b, and returns local velocity

These conversion functions can live in `jetson_nano/services/motor.h/cpp`. 

### Blocked By

N/A

### PR DESCRIPTION
**Pull request by liambontkes**
### Description

Implemented a Euclidean velocity to wheel speed conversion class. The class takes into account the robot dynamics to perform accurate wheel speed calculations. 

### Testing Done

All directions verified with cross check in MATLAB. MATLAB script is part of the PR.

### Resolved Issues

resolves #2406 

### Length Justification and Key Files to Review

Key files: physics.h, physics.cpp

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/physics/physics.cpp [hunk 1]
```diff
@@ -25,3 +29,133 @@ Vector calculateFutureVelocity(const Vector &initial_velocity,
     double vy1 = initial_velocity.y() + seconds_in_future * constant_acceleration.y();
     return Vector(vx1, vy1);
 }
+
+EuclideanToFourWheel::EuclideanToFourWheel(double front_wheel_angle_deg,
+                                           double back_wheel_angle_deg)
+{
+    auto robot_constants = create2021RobotConstants();
```
Comment (akhilveeraghanta) [inline src/software/physics/physics.cpp]: you can take a RobotConstants_t struct as an input here

File: src/software/physics/physics.cpp [hunk 2]
```diff
@@ -25,3 +29,133 @@ Vector calculateFutureVelocity(const Vector &initial_velocity,
...
+    auto j = cos(front_wheel_angle_phi_) / (2 * pow(cos(front_wheel_angle_phi_), 2) +
+                                            2 * pow(cos(front_wheel_angle_phi_), 2));
+    auto k = sin(rear_wheel_angle_theta_) /
+             (2 * sin(front_wheel_angle_phi_), 2 * sin(rear_wheel_angle_theta_));
+
+    wheel_speed_to_euclidean_velocity_D_inverse_ << -i, -i, i, i, j, -j, -(1. - j),
+        (1. - j), k, k, (1 - k), (1 - k);
+}
+
+WheelSpace_t EuclideanToFourWheel::get_target_wheel_speeds(
```
Comment (akhilveeraghanta) [inline src/software/physics/physics.cpp]: we use `camelCase` for our function names and  `snake_case` for our variables

variables look good, but the functions need to be reformatted to `snakeCase` i.e `getTargetWheelSpeeds`

File: src/software/physics/physics.cpp [hunk 3]
```diff
@@ -25,3 +29,133 @@ Vector calculateFutureVelocity(const Vector &initial_velocity,
...
+    auto robot_constants = create2021RobotConstants();
+
+    // TODO: replace with Thunderloop polling rate constant
+    // Set to period of control loop.
+    delta_t_ = 1.0 / 200;
+
+    // import robot constants
+    robot_mass_M_ = robot_constants.mass_kg;
+    // TODO: update
+    robot_radius_R_ = 1;
```
Comment (akhilveeraghanta) [inline src/software/physics/physics.cpp]: you can use `ROBOT_MAX_RADIUS` here

File: src/software/physics/physics.h [hunk 4]
```diff
@@ -36,3 +38,152 @@ Point calculateFuturePosition(const Point &initial_position,
...
+typedef Eigen::Vector3d EuclideanSpace_t;
+
+/**
+ * Vector representation of the robot wheel space.
+ *
+ * 0: Front Right
+ * 1: Back Right
+ * 2: Back Left
+ * 3: Front Left
+ * TODO: Verify wheel orientations
```
Comment (akhilveeraghanta) [inline src/software/physics/physics.h]: lets pick the indexing thats consistent with the omindrive code
ill update the motor code to reflect that as well

**Review by akhilveeraghanta — COMMENTED** [review]
did an initial pass, looks great! can't wait to try it

File: src/software/physics/physics.cpp [hunk 5]
```diff
@@ -25,3 +32,123 @@ Vector calculateFutureVelocity(const Vector &initial_velocity,
...
+    // import robot constants
+    robot_mass_M_            = robot_constants.mass_kg;
+    robot_radius_R_          = robot_constants.robot_radius_m;
+    inertial_factor_alpha_   = robot_constants.inertial_factor;
+    front_wheel_angle_phi_   = robot_constants.front_wheel_angle_deg * M_PI / 180.;
+    rear_wheel_angle_theta_  = (robot_constants.back_wheel_angle_deg - 90.) * M_PI / 180.;
+
+    // calculate coupling matrices
+    // ref: http://robocup.mi.fu-berlin.de/buch/omnidrive.pdf
+    auto a = pow(sin(front_wheel_angle_phi_), 2) - pow(cos(front_wheel_angle_phi_), 2);
```
Comment (akhilveeraghanta) [inline src/software/physics/physics.cpp]: its worth mentioning here that a,b,c,d,e,f,i,j,k all correspond to page 17 of that document

File: src/software/physics/physics.cpp [hunk 6]
```diff
@@ -25,3 +32,123 @@ Vector calculateFutureVelocity(const Vector &initial_velocity,
...
+    const EuclideanSpace_t &initial_velocity,
+    const EuclideanSpace_t &target_velocity) const
+{
+    return (target_velocity - initial_velocity) / delta_t_;
+}
+
+WheelSpace_t EuclideanToFourWheel::getTranslationalWheelForces(
+    EuclideanSpace_t target_acceleration) const
+{
+    // x acceleration
```
Comment (akhilveeraghanta) [inline src/software/physics/physics.cpp]: its worth mentioning these calculations are from page 14

File: src/software/physics/physics.cpp [hunk 7]
```diff
@@ -25,3 +32,123 @@ Vector calculateFutureVelocity(const Vector &initial_velocity,
     double vy1 = initial_velocity.y() + seconds_in_future * constant_acceleration.y();
     return Vector(vx1, vy1);
 }
+
+EuclideanToFourWheel::EuclideanToFourWheel()
```
Comment (akhilveeraghanta) [inline src/software/physics/physics.cpp]: this class is important enough that I think it warrants its own file, id recommend moving this to 

`eucliedean_to_four_wheel.h` and `eucliedean_to_four_wheel.cpp`

File: src/firmware/boards/powerboard_esp32/BUILD [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package(default_visibility = ["//visibility:public"])
```
Comment (akhilveeraghanta) [inline src/firmware/boards/powerboard_esp32/BUILD]: you can remove all the powerboard files

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
looks great! just a couple nits

File: src/software/physics/physics_test.cpp [hunk 9]
```diff
@@ -99,3 +101,118 @@ TEST(PhysicsUtilTest, test_future_velocity_sideways_acceleration)
                                 Duration::fromSeconds(2.5)),
         0.001));
 }
+
+class EuclideanToFourWheelTest : public ::testing::Test
```
Comment (itsarune) [inline src/software/physics/physics_test.cpp]: this could be a separate test file as well

**Review by itsarune — COMMENTED** [review]
looks good after akhil's suggestions + those build issues

File: src/matlab_tools/euclidean_to_wheel.m [hunk 10]
```diff
@@ -0,0 +1,56 @@
+function target_wheel_speeds = euclidean_to_wheel(current_wheel_speeds, target_euclidean_velocity)
```
Comment (akhilveeraghanta) [inline src/matlab_tools/euclidean_to_wheel.m]: nit: maybe put a note up top here and say this is a test script that can be used to check the real thing

you can also move this file to `src/physics`, i don't think we will have any more matlab tools in this folder

File: src/software/physics/physics.cpp [hunk 11]
```diff
@@ -1,5 +1,12 @@
 #include "software/physics/physics.h"
 
+#define _USE_MATH_DEFINES
+
+#include <cmath>
+
+#include "shared/2021_robot_constants.h"
+#include "shared/constants.h"
+
```
Comment (akhilveeraghanta) [inline src/software/physics/physics.cpp]: nit: can remove

File: src/software/physics/physics.h [hunk 12]
```diff
@@ -1,5 +1,7 @@
 #pragma once
 
+#include <Eigen/Dense>
+
```
Comment (akhilveeraghanta) [inline src/software/physics/physics.h]: nit: can remove

File: src/software/physics/physics.cpp (hunk 11 above)
Comment (akhilveeraghanta) [reply src/software/physics/physics.cpp]: this is what seems to be causing CI to fail too

**Review by akhilveeraghanta — COMMENTED** [review]
looks good on my end, just a couple final comments on fixing CI

think you also need to run `./formatting_scripts/fix_formatting.sh`

File: src/matlab_tools/euclidean_to_wheel.m (hunk 10 above)
Comment (liambontkes) [reply src/matlab_tools/euclidean_to_wheel.m]: do you mean create a new folder `src/physics` or move it to `src/software/physics`

Comment (akhilveeraghanta) [reply src/matlab_tools/euclidean_to_wheel.m]: ah my bad, I mean't `src/software/physsics` same folder as the other class

**Review by akhilveeraghanta — DISMISSED** [review]
lgtm! nice work, can't wait to try it

File: src/shared/2021_robot_constants.c [hunk 13]
```diff
@@ -22,6 +22,9 @@ extern "C"
     {
         RobotConstants_t robot_constants = {
             .mass_kg = 2.5f,  // determined experimentally
+            // TODO: update units
```
Comment (jonathanlew) [inline src/shared/2021_robot_constants.c]: TODOs should be tagged with an issue or removed (arguably inertial factor is unitless)

File: src/software/physics/euclidean_to_wheel.cpp [hunk 14]
```diff
@@ -0,0 +1,128 @@
+#include "euclidean_to_wheel.h"
+
+#define _USE_MATH_DEFINES
```
Comment (jonathanlew) [inline src/software/physics/euclidean_to_wheel.cpp]: why's this needed?

File: src/software/physics/euclidean_to_wheel.h [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+    EuclideanToWheel();
+
+    /**
+     * Gets wheel velocity targets from the desired Euclidean velocity.
+     *
+     * Step 1: Convert Euclidean velocity to acceleration.
+     * Step 2: Calculate the target translational wheel forces.
+     * Step 3: Calculate the target rotational wheel forces.
+     * Step 4: Sum the wheel forces.
+     * Step 5: Convert wheel forces to speeds.
```
Comment (jonathanlew) [inline src/software/physics/euclidean_to_wheel.h]: implementation details should be in the source file (.cpp)

File: src/software/physics/euclidean_to_wheel.h [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+     * @return The equivalent wheel speeds.
+     */
+    WheelSpace_t getTargetWheelSpeeds(const EuclideanSpace_t &target_euclidean_velocity,
+                                      const WheelSpace_t &current_wheel_speeds);
+
+   private:
+    /**
+     * The control loop time period [s].
+     */
+    double delta_t_{};
```
Comment (jonathanlew) [inline src/software/physics/euclidean_to_wheel.h]: should probably put units in all these field names

File: src/software/physics/euclidean_to_wheel.m [hunk 17]
```diff
@@ -0,0 +1,56 @@
+function target_wheel_speeds = euclidean_to_wheel(current_wheel_speeds, target_euclidean_velocity)
+%EUCLIDEAN_TO_WHEEL Converts Euclidean velocity to robot wheel speeds.
```
Comment (jonathanlew) [inline src/software/physics/euclidean_to_wheel.m:2]: curious what this file is all about

File: src/software/physics/euclidean_to_wheel.cpp [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+#include "shared/2021_robot_constants.h"
+
+EuclideanToWheel::EuclideanToWheel()
+{
+    auto robot_constants = create2021RobotConstants();
+
+    // TODO: replace with Thunderloop polling rate constant
+    // Set to period of control loop.
+    delta_t_ = 1. / 200;
+
```
Comment (jonathanlew) [inline src/software/physics/euclidean_to_wheel.cpp]: this should be passed in as a constructor argument

**Review by jonathanlew — CHANGES_REQUESTED** [review]
cool that we're going have much better wheel conversion, just a few minor comments

File: src/software/physics/euclidean_to_wheel.m (hunk 17 above)
Comment (liambontkes) [reply src/software/physics/euclidean_to_wheel.m:2]: This was for cross checking the formulas, just did it through matlab.

File: src/software/physics/euclidean_to_wheel.cpp (hunk 14 above)
Comment (liambontkes) [reply src/software/physics/euclidean_to_wheel.cpp]: Lines 21, 22 its used to convert degrees to radians.

File: src/software/physics/euclidean_to_wheel.h [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+ * 1: Front Left
+ * 2: Back Left
+ * 3: Back Right
+ */
+typedef Eigen::Vector4d WheelSpace_t;
+
+class EuclideanToWheel
+{
+   public:
+    EuclideanToWheel() = default;
```
Comment (jonathanlew) [inline src/software/physics/euclidean_to_wheel.h]: nit: we only want to construct this class with the control_loop_frequency_Hz
```suggestion
    EuclideanToWheel() = delete;
```

File: src/software/physics/euclidean_to_wheel_test.cpp [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+
+    static void SetUp() override
+    {
+        auto euclideanToFourWheel = new EuclideanToWheel(200);
+    }
+
+    static void TearDown() override
+    {
+        delete euclideanToFourWheel;
+    }
```
Comment (jonathanlew) [inline src/software/physics/euclidean_to_wheel_test.cpp]: do this and replace euclideanToFourWheel with euclidean_to_four_wheel in this file to fix the build error
```suggestion
    EuclideanToWheel euclidean_to_four_wheel = EuclideanToWheel(200.0f);
```

**Review by jonathanlew — COMMENTED** [review]
just a couple more comments from me

**Review by akhilveeraghanta — APPROVED** [review]
lgtm!

**Review by jonathanlew — APPROVED** [review]

**Pull request merged**